### ISSUE TITLE
Custom Attention Mask

### ISSUE THREAD

**Issue by PaulCouairon** [SHI-Labs/NATTEN#249]
Hi, 

I am trying to use a custom attention mask (for example mask scores below a threshold). The flex-attention backend seems to be suited for this but I am not fully sure to understand how to pass a ```score_mod``` function as an argument. 

In addition, is #231 compatible with flex-attention backend? 

Thank you in advance!

**Comment by alihassanijr** [comment]
>for example masking out scores below a certain threshold

If it's just a single floating point that clips the dot products, I think it might be easy enough to add to our CUDA backends as well -- but again whether it ends up being a permanent feature is a matter of how common / important the use case is.

**Comment by PaulCouairon** [comment]
Yes, it’s really just a single floating-point threshold that clips the dot products, similar to the masking described in [ProxyCLIP](https://arxiv.org/abs/2408.04883). I understand your point about whether it’s worth making permanent.

**Comment by alihassanijr** [comment]
I think that should be easy enough to implement in all backends.
I'll look into this.

**Comment by alihassanijr** [comment]
Just to clarify though, we're interested in masking the dot products, not attention weights, correct?

Something like:

```python3
# dot_product_min_value: float

S = q @ k.T
S = S.clip(min=dot_product_min_value)
A = S.softmax(dim=-1)
output = A @ v

```

**Comment by PaulCouairon** [comment]
Yes exactly!

**Comment by PaulCouairon** [comment]
I am using Ampere architectures (A6000, A100).

**Comment by alihassanijr** [comment]
I assumed clip then scale for now, but it's easy enough to change.

You can build it off of this branch:

https://github.com/alihassanijr/NATTEN-Torch/tree/feat/dot-product-clipping


PR to track: #252


NOTE: we don't have gradchecks in NATTEN anymore, and since this feature doesn't exist in SDPA or other standard attention APIs / kernels (as far as I know), I strongly recommend verifying this against some reference implementation in pure PyTorch.
You don't need to implement neighborhood attention -- you can just set `kernel_size` to your input size, and not do dilation, stride, or causal masking, and then compare against a pure PyTorch implementation of attention with dot product clipping.

I'll try to do that myself too, but I might not have time to work on it this week.

**Comment by PaulCouairon** [comment]
Thank you very much for your help! I'll try this.

**Comment by alihassanijr** [comment]
I'll take a look this weekend.

**Comment by alihassanijr** [comment]
I still haven't been able to get it to pass our entire sweep with the same error thresholds.
I'm still unsure whether this is due to a bug I'm not spotting, or whether it's reasonable for this functionality to be so sensitive to small accum differences between the reference and target kernels.

I suggest you test it in whatever harness you might have, but please be cautious when using this in training. Even if it's not a bug, this issue could highlight a potential shortcoming of this approach.

https://github.com/alihassanijr/NATTEN-Torch/tree/feat/dot-product-clipping

**Comment by PaulCouairon** [comment]
Thank you very much again! I'll try this!
